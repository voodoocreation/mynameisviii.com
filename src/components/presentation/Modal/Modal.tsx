import cn from "classnames";
import * as React from "react";
import * as ReactDOM from "react-dom";
import { MdClose } from "react-icons/lib/md";

import { lockScroll, unlockScroll } from "../../../helpers/dom";

interface IProps {
  className?: string;
  isOpen?: boolean;
  onClose: () => void;
  onKeyDown?: (event: KeyboardEvent) => void;
  hasFocusRestriction?: boolean;
  hasOverlayClick?: boolean;
}

interface IState {
  isVisible: boolean;
}

export default class Modal extends React.Component<IProps, IState> {
  public static defaultProps = {
    hasFocusRestriction: true,
    hasOverlayClick: true
  };

  public readonly state = {
    isVisible: false
  };

  private container: HTMLDivElement | null = null;
  private modalRef: React.RefObject<HTMLDivElement> = React.createRef();

  public componentDidMount() {
    this.container = document.querySelector(".Portal");

    if (this.props.isOpen) {
      this.onShow();
    }
  }

  public componentDidUpdate(prevProps: IProps) {
    if (prevProps.isOpen && !this.props.isOpen) {
      this.onHide();
    } else if (!prevProps.isOpen && this.props.isOpen) {
      this.onShow();
    }
  }

  public componentWillUnmount() {
    if (this.props.isOpen) {
      this.onHide();
    }
  }

  public render() {
    return this.container && this.props.isOpen
      ? ReactDOM.createPortal(this.renderModal(), this.container)
      : null;
  }

  private renderModal() {
    const { children, className, hasOverlayClick, isOpen } = this.props;
    const { isVisible } = this.state;

    return (
      <div
        aria-hidden={!isOpen}
        aria-modal={true}
        className={cn("Modal", className, { isVisible })}
        ref={this.modalRef}
        role="dialog"
        tabIndex={isOpen ? 0 : -1}
      >
        <div
          className="Modal-overlay"
          onClick={hasOverlayClick ? this.onClose : undefined}
        />

        <div className="Modal-content">
          <div className="Modal-body">{children}</div>

          <button className="Modal-closeButton" onClick={this.onClose}>
            <MdClose />
          </button>
        </div>
      </div>
    );
  }

  private onFocus = (event: FocusEvent) => {
    if (
      this.container &&
      event.target &&
      this.props.isOpen &&
      this.props.hasFocusRestriction
    ) {
      if (
        event.target === window ||
        !this.container.contains(event.target as Node)
      ) {
        event.stopPropagation();

        if (this.modalRef.current) {
          this.modalRef.current.focus();
        }
      }
    }
  };

  private onClose = () => {
    this.props.onClose();
  };

  private onKeyDown = (event: KeyboardEvent) => {
    if (event.key === "Escape") {
      this.onClose();
    }

    if (this.props.onKeyDown) {
      this.props.onKeyDown(event);
    }
  };

  private onShow = () => {
    if (this.container) {
      if (!this.state.isVisible) {
        setTimeout(() => {
          this.setState({
            isVisible: true
          });
        }, 1);
      }

      if (this.props.hasFocusRestriction) {
        window.addEventListener("focus", this.onFocus, true);
      }

      window.addEventListener("keydown", this.onKeyDown);

      lockScroll();

      if (this.modalRef.current) {
        this.modalRef.current.focus();
      }
    }
  };

  private onHide = () => {
    if (this.container) {
      if (this.state.isVisible) {
        this.setState({
          isVisible: false
        });
      }

      if (this.props.hasFocusRestriction) {
        window.removeEventListener("focus", this.onFocus, true);
      }

      window.removeEventListener("keypress", this.onKeyDown);

      unlockScroll();
    }
  };
}
