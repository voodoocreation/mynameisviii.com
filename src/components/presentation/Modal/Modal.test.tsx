import { filterCalls } from "jest-mocks";
import * as React from "react";

import WrapperWithIntl from "../../../utilities/WrapperWithIntl";
import Modal from "./Modal";

const component = new WrapperWithIntl(Modal)
  .withDefaultProps({
    className: "TestModal",
    hasFocusRestriction: true,
    isOpen: true,
    onClose: jest.fn(),
    usePortal: false
  })
  .withDefaultChildren(
    <div className="Modal--testContent">
      <input id="Modal--input" />
    </div>
  );

describe("[presentation] <Modal />", () => {
  jest.spyOn(window, "addEventListener");
  jest.spyOn(window, "removeEventListener");

  it("doesn't render anything when isOpen is false", () => {
    const wrapper = component.withProps({ isOpen: false }).mount();

    expect(wrapper.render().html()).toBeNull();
  });

  describe("when toggling the isOpen prop", () => {
    let wrapper: ReturnType<typeof component.mount>;

    beforeAll(() => {
      jest.clearAllMocks();
    });

    afterAll(() => {
      wrapper.unmount();
    });

    it("mounts the component", () => {
      wrapper = component.mount();
    });

    it("renders the modal", () => {
      expect(wrapper.find(".Modal")).toHaveLength(1);
    });

    it("matches snapshot when open", () => {
      expect(wrapper.render()).toMatchSnapshot();
    });

    it("renders the children in the body element", () => {
      expect(wrapper.find(".Modal--body .Modal--testContent")).toHaveLength(1);
    });

    it("sets the isOpen prop to false", () => {
      wrapper.setProps({ isOpen: false });
    });

    it("doesn't render anything", () => {
      expect(wrapper.render().html()).toBeNull();
    });

    it("sets the isOpen prop to true again", () => {
      wrapper.setProps({ isOpen: true });
    });

    it("renders the modal", () => {
      expect(wrapper.find(".Modal")).toHaveLength(1);
    });

    it("sets the isOpen prop to false", () => {
      wrapper.setProps({ isOpen: false });
    });
  });

  describe("when interacting with the close button", () => {
    let wrapper: ReturnType<typeof component.mount>;

    beforeAll(() => {
      jest.clearAllMocks();
    });

    afterAll(() => {
      wrapper.unmount();
    });

    it("mounts the component", () => {
      wrapper = component.mount();
    });

    it("clicks the close button", () => {
      wrapper.find("Button.Modal--closeButton").simulate("click");
    });

    it("calls the onClose prop", () => {
      expect(component.props.onClose).toHaveBeenCalledTimes(1);
    });
  });

  describe("when interacting with the overlay", () => {
    let wrapper: ReturnType<typeof component.mount>;

    beforeAll(() => {
      jest.clearAllMocks();
    });

    afterAll(() => {
      wrapper.unmount();
    });

    it("mounts the component", () => {
      wrapper = component.mount();
    });

    it("clicks the overlay", () => {
      wrapper.find(".Modal--overlay").simulate("click");
    });

    it("calls the onClose prop", () => {
      expect(component.props.onClose).toHaveBeenCalledTimes(1);
    });

    it("sets the hasOverlayClick prop to false", () => {
      wrapper.setProps({ hasOverlayClick: false });
    });

    it("clicks the overlay", () => {
      wrapper.find(".Modal--overlay").simulate("click");
    });

    it("doesn't call the onClose prop", () => {
      expect(component.props.onClose).toHaveBeenCalledTimes(1);
    });
  });

  describe("when interacting with the keyboard", () => {
    let wrapper: ReturnType<typeof component.mount>;

    beforeAll(() => {
      jest.clearAllMocks();
    });

    it("mounts the component", () => {
      wrapper = component
        .withProps({
          onKeyDown: jest.fn()
        })
        .mount();
    });

    it("binds keydown event handler", () => {
      expect(
        filterCalls(window.addEventListener as jest.Mock, "keydown")
      ).toHaveLength(1);
    });

    it("presses the Escape key", () => {
      window.dispatchEvent(new KeyboardEvent("keydown", { key: "Escape" }));
    });

    it("calls the onClose prop", () => {
      expect(component.props.onClose).toHaveBeenCalledTimes(1);
    });

    it("calls the onKeyDown prop", () => {
      expect(component.props.onKeyDown).toHaveBeenCalledTimes(1);
    });

    it("presses a key other than Escape", () => {
      window.dispatchEvent(new KeyboardEvent("keydown", { key: "ArrowLeft" }));
    });

    it("doesn't call the onClose prop", () => {
      expect(component.props.onClose).toHaveBeenCalledTimes(1);
    });

    it("calls the onKeyDown prop", () => {
      expect(component.props.onKeyDown).toHaveBeenCalledTimes(2);
    });

    it("sets the onKeyDown prop to be undefined", () => {
      wrapper.setProps({ onKeyDown: undefined });
    });

    it("presses the Escape key", () => {
      window.dispatchEvent(new KeyboardEvent("keydown", { key: "Escape" }));
    });

    it("calls the onClose prop", () => {
      expect(component.props.onClose).toHaveBeenCalledTimes(2);
    });

    it("sets the isOpen prop to false", () => {
      wrapper.setProps({ isOpen: false });
    });

    it("unbinds keydown event handler", () => {
      expect(
        filterCalls(window.removeEventListener as jest.Mock, "keydown")
      ).toHaveLength(1);
    });

    it("sets the isOpen prop to true", () => {
      wrapper.setProps({ isOpen: true });
    });

    it("binds keydown event handler", () => {
      expect(
        filterCalls(window.addEventListener as jest.Mock, "keydown")
      ).toHaveLength(2);
    });

    it("unmounts component", () => {
      wrapper.unmount();
    });

    it("unbinds keydown event handler", () => {
      expect(
        filterCalls(window.removeEventListener as jest.Mock, "keydown")
      ).toHaveLength(2);
    });
  });

  describe("Focus management", () => {
    describe("when binding the focus event handler", () => {
      let wrapper: ReturnType<typeof component.mount>;

      beforeAll(() => {
        jest.clearAllMocks();
      });

      afterAll(() => {
        wrapper.unmount();
      });

      it("mounts the component", () => {
        wrapper = component
          .withProps({
            hasFocusRestriction: true
          })
          .mount();
      });

      it("binds focus event handler", () => {
        expect(
          filterCalls(window.addEventListener as jest.Mock, "focus")
        ).toHaveLength(1);
      });

      it("sets the isOpen prop to false", () => {
        wrapper.setProps({ isOpen: false });
      });

      it("unbinds focus event handler", () => {
        expect(
          filterCalls(window.removeEventListener as jest.Mock, "focus")
        ).toHaveLength(1);
      });

      it("sets the isOpen prop to true", () => {
        wrapper.setProps({ isOpen: true });
      });

      it("binds focus event handler", () => {
        expect(
          filterCalls(window.addEventListener as jest.Mock, "focus")
        ).toHaveLength(2);
      });

      it("sets the hasFocusRestriction prop to false", () => {
        wrapper.setProps({ hasFocusRestriction: false });
      });

      it("unbinds focus event handler", () => {
        expect(
          filterCalls(window.removeEventListener as jest.Mock, "focus")
        ).toHaveLength(2);
      });

      it("sets the isOpen prop to false", () => {
        wrapper.setProps({ isOpen: false });
      });

      it("doesn't unbind the focus event handler", () => {
        expect(
          filterCalls(window.removeEventListener as jest.Mock, "focus")
        ).toHaveLength(2);
      });

      it("sets the isOpen prop to true", () => {
        wrapper.setProps({ isOpen: true });
      });

      it("doesn't bind the focus event handler", () => {
        expect(
          filterCalls(window.removeEventListener as jest.Mock, "focus")
        ).toHaveLength(2);
      });

      it("sets the hasFocusRestriction prop to true", () => {
        wrapper.setProps({ hasFocusRestriction: true });
      });

      it("binds focus event handler", () => {
        expect(
          filterCalls(window.addEventListener as jest.Mock, "focus")
        ).toHaveLength(3);
      });
    });

    describe("when dispatching focus events", () => {
      const stopPropagationMock = jest.fn();

      const bodyEvent = new FocusEvent("focus");
      Object.defineProperties(bodyEvent, {
        stopPropagation: { value: stopPropagationMock },
        target: { value: document.body }
      });

      const inputNode = document.createElement("input");
      document.body.appendChild(inputNode);

      const outsideInputEvent = new FocusEvent("focus");
      Object.defineProperties(outsideInputEvent, {
        stopPropagation: { value: stopPropagationMock },
        target: { value: inputNode }
      });

      describe("when usePortal is false", () => {
        let wrapper: ReturnType<typeof component.mount>;

        beforeAll(() => {
          jest.clearAllMocks();
        });

        afterAll(() => {
          wrapper.unmount();
        });

        it("mounts the component", () => {
          wrapper = component.mount();
        });

        it("dispatches a focus event from the body element", () => {
          window.dispatchEvent(bodyEvent);
        });

        it("prevents the focus event from proceeding", () => {
          expect(stopPropagationMock).toHaveBeenCalledTimes(1);
        });

        it("dispatches a focus event from an input element outside the modal", () => {
          window.dispatchEvent(outsideInputEvent);
        });

        it("prevents the focus event from proceeding", () => {
          expect(stopPropagationMock).toHaveBeenCalledTimes(2);
        });

        it("dispatches a focus event from an input element within the modal", () => {
          const modalInput = wrapper.find("#Modal--input").instance();
          const event = new FocusEvent("focus");
          Object.defineProperties(event, {
            stopPropagation: { value: stopPropagationMock },
            target: { value: modalInput }
          });
          window.dispatchEvent(event);
        });

        it("doesn't prevent the focus event from proceeding", () => {
          expect(stopPropagationMock).toHaveBeenCalledTimes(2);
        });

        it("clears stopPropagationMock", () => {
          stopPropagationMock.mockClear();
        });
      });

      describe("when usePortal is true", () => {
        let wrapper: ReturnType<typeof component.mount>;

        beforeAll(() => {
          jest.clearAllMocks();
        });

        afterAll(() => {
          wrapper.unmount();
        });

        it("mounts the component", () => {
          wrapper = component
            .withProps({
              usePortal: true
            })
            .mount();
        });

        it("dispatches a focus event from the body element", () => {
          window.dispatchEvent(bodyEvent);
        });

        it("prevents the focus event from proceeding", () => {
          expect(stopPropagationMock).toHaveBeenCalledTimes(1);
        });

        it("dispatches a focus event from an input element outside the modal", () => {
          window.dispatchEvent(outsideInputEvent);
        });

        it("prevents the focus event from proceeding", () => {
          expect(stopPropagationMock).toHaveBeenCalledTimes(2);
        });

        it("dispatches a focus event from an input element within the modal", () => {
          const modalInput = wrapper.find("#Modal--input").instance();
          const event = new FocusEvent("focus");
          Object.defineProperties(event, {
            stopPropagation: { value: stopPropagationMock },
            target: { value: modalInput }
          });
          window.dispatchEvent(event);
        });

        it("doesn't prevent the focus event from proceeding", () => {
          expect(stopPropagationMock).toHaveBeenCalledTimes(2);
        });

        it("clears stopPropagationMock", () => {
          stopPropagationMock.mockClear();
        });
      });
    });
  });
});
