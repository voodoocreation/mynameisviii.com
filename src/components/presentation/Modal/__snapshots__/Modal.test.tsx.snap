// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[presentation] <Modal /> when toggling the isOpen prop matches snapshot when open 1`] = `
<div
  class="Portal--inPlace"
>
  <div
    aria-hidden="false"
    aria-modal="true"
    class="Modal TestModal isVisible"
    role="dialog"
  >
    <button
      class="Button Modal--overlay"
      type="button"
    >
      <span
        class="Button--content"
      />
    </button>
    <div
      class="Modal--content"
    >
      <div
        class="Modal--body"
      >
        <div
          class="Modal--testContent"
        >
          <input
            id="Modal--input"
          />
        </div>
      </div>
      <button
        class="Button Modal--closeButton"
        type="button"
      >
        <span
          class="Button--content"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
        </span>
      </button>
    </div>
  </div>
</div>
`;
